### PR TITLE
Change how cover accessories are handled

### PR DESCRIPTION
**Pull request by bSr43**
Hello,
This is a very raw attempt at fixing the issue I was exposing earlier today.
Please keep in mind that I've never touched any TypeScript project before, nor any tools like npm (at least for publishing), so my guess is that the result is far from perfect; I usually write C and ASM for a living 😅
Anyway, here is the thing: the status of the cover cannot be retrieved from the cloud. Hence, I'm keeping a few state variables which try to reflect the status. The `PositionState` and `CurrentPosition` characteristics are only wrapper around those variables. The `TargetPosition` is the one which trigger the different characteristics in sequence.
So far, it seems to work on my setup… 😊

### CODE REVIEW

**Comment by iDreamsX** [comment]
Hello @bSr43, 

Thank you for this correction on the shutters. It works correctly now.

But I notice a problem, when they are opened manually (with the switch), the information is not up and does not update on Homekit.

I set the interval time to 800 seconds, but it doesn't update.

Thank you for this correction which works!

**Comment by bSr43** [comment]
Unfortunately, there is no way to know if the button has been pressed manually, and in that case it ends up in a loss of synchronization with HomeKit 😕 

Anyway, happy to see that it (almost) works for you 😊 
In my case, I’ve switched to Shelly 2.5 with works way better than those switches...

**Comment by iDreamsX** [comment]
Too bad, I was hoping there was a way to look at the status of the last effective action or state. 
Let's hope that during a next update we'll be able to look at that to improve the plugin!

**Comment by milo526** [comment]
> Too bad, I was hoping there was a way to look at the status of the last effective action or state.
> Let's hope that during a next update we'll be able to look at that to improve the plugin!

The Tuya API that this plugin uses does not support this.
This is not possible and not something I or anybody outside of tuya can fix.

**Comment by rulez86** [comment]
Would be really nice to merge fork from bSr43 to make covers working. Any news when for is ready to use?

File: src/accessories/CoverAccessory.ts [hunk 1]
```diff
@@ -10,6 +10,10 @@ import { BaseAccessory } from "./BaseAccessory";
 import { TuyaDevice } from "../api/response";
 
 export class CoverAccessory extends BaseAccessory {
+  public target = 0;
```
Comment (milo526) [inline src/accessories/CoverAccessory.ts:13]: Please change these [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) to enums

File: src/accessories/CoverAccessory.ts [hunk 2]
```diff
@@ -10,6 +10,10 @@ import { BaseAccessory } from "./BaseAccessory";
...
+  public position = 0;
```
Comment (milo526) [inline src/accessories/CoverAccessory.ts:14]: Please change these [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) to enums

File: src/accessories/CoverAccessory.ts [hunk 3]
```diff
@@ -10,6 +10,10 @@ import { BaseAccessory } from "./BaseAccessory";
...
+  public motor = 2;
```
Comment (milo526) [inline src/accessories/CoverAccessory.ts:15]: Please change these [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) to enums

That will also possibly explain why this value defaults to 2 and the others default to 0

File: src/accessories/characteristics/currentPosition.ts [hunk 4]
```diff
@@ -15,43 +16,10 @@ export class CurrentPositionCharacteristic extends TuyaWebCharacteristic {
...
 
   public getRemoteValue(callback: CharacteristicGetCallback): void {
-    this.accessory
-      .getDeviceState()
-      .then((data) => {
-        this.debug("[GET] %s", data?.state);
-        this.updateValue(data, callback);
-      })
-      .catch(this.accessory.handleError("GET", callback));
+    this.updateValue({}, callback);
```
Comment (milo526) [inline src/accessories/characteristics/currentPosition.ts:19]: Please revert the changes here so that remove changes are reflect properly i.e. the state is opening or closing (i.e. motor active) when the API is currently reporting this state.

File: src/accessories/characteristics/currentPosition.ts [hunk 5]
```diff
@@ -15,43 +16,10 @@ export class CurrentPositionCharacteristic extends TuyaWebCharacteristic {
...
-        this.homekitCharacteristic,
-        stateValue,
-        !callback
-      );
-      callback && callback(null, stateValue);
-    } else {
-      callback &&
-        callback(new Error(`Unexpected state value provided: ${data?.state}`));
-    }
+    callback && callback(null, (<CoverAccessory>this.accessory).position);
```
Comment (milo526) [inline src/accessories/characteristics/currentPosition.ts:23]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/positionState.ts [hunk 6]
```diff
@@ -14,21 +15,11 @@ export class PositionStateCharacteristic extends TuyaWebCharacteristic {
...
 
   updateValue(data: DeviceState, callback?: CharacteristicGetCallback): void {
-    this.debug("Setting position state to stopped");
-    this.accessory.setCharacteristic(
-      this.homekitCharacteristic,
-      this.PositionState.STOPPED,
-      !callback
-    );
-    callback && callback(null, this.PositionState.STOPPED);
+    callback && callback(null, (<CoverAccessory>this.accessory).motor);
```
Comment (milo526) [inline src/accessories/characteristics/positionState.ts:23]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/targetPosition.ts [hunk 7]
```diff
@@ -33,13 +34,8 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
 
   public getRemoteValue(callback: CharacteristicGetCallback): void {
-    this.accessory
-      .getDeviceState()
-      .then((data) => {
-        this.debug("[GET] %s", data?.state);
-        this.updateValue(data, callback);
-      })
-      .catch(this.accessory.handleError("GET", callback));
+    const a = <CoverAccessory>this.accessory;
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:37]: Please in-line this variable into the next line as you also do in the other files

File: src/accessories/characteristics/targetPosition.ts [hunk 8]
```diff
@@ -33,13 +34,8 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
+    callback && callback(null, a.target);
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:38]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/targetPosition.ts [hunk 9]
```diff
@@ -48,50 +44,59 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
+      .setDeviceState("turnOnOff", { value }, value)
       .then(async () => {
-        this.debug("[SET] %s", value);
+        this.debug("[SET] turnOnOff command sent with value %s", value);
         callback();
 
-        await delay(1000);
-        this.accessory.setTuyaCharacteristic(
+        this.debug("Setting targetPosition to %d", target);
+        coverAccessory.target = target;
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:59]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/targetPosition.ts [hunk 10]
```diff
@@ -48,50 +44,59 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
+
+        coverAccessory.position = target;
+        this.accessory.setCharacteristic(
           this.accessory.platform.Characteristic.CurrentPosition,
-          data
+          coverAccessory.position,
+          true
+        );
+
+        coverAccessory.motor = this.accessory.platform.Characteristic.PositionState.STOPPED;
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:89]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/targetPosition.ts [hunk 11]
```diff
@@ -48,50 +44,59 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
+        );
+
+        await delay(5000);
+
+        this.debug(
+          "Setting currentPosition to %d and positionState to STOPPED",
+          target
+        );
+
+        coverAccessory.position = target;
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:82]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/targetPosition.ts [hunk 12]
```diff
@@ -48,50 +44,59 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
+          target,
+          true
+        );
+
+        coverAccessory.motor = value
+          ? this.accessory.platform.Characteristic.PositionState.INCREASING
+          : this.accessory.platform.Characteristic.PositionState.DECREASING;
+        this.accessory.setCharacteristic(
+          this.accessory.platform.Characteristic.PositionState,
+          coverAccessory.motor,
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:71]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

File: src/accessories/characteristics/targetPosition.ts [hunk 13]
```diff
@@ -48,50 +44,59 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
...
-        this.homekitCharacteristic,
-        stateValue,
-        !callback
-      );
-      callback && callback(null, stateValue);
-    } else {
-      callback &&
-        callback(new Error(`Unexpected state value provided: ${data?.state}`));
-    }
+    callback && callback(null, (<CoverAccessory>this.accessory).target);
```
Comment (milo526) [inline src/accessories/characteristics/targetPosition.ts:100]: This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic

**Review by milo526 — CHANGES_REQUESTED** [review]

**Comment by milo526** [comment]
Thank you for this MR once more.
To answer @rulez86 question, there are a few changes that I'd like to see changed.
There is also a merge-conflict that needs to be solved.

@bSr43 are you able and willing to fix this?

Since I do not own a Tuya Cover accessory, it is hard for me to test this and feature development will be slow.

**Comment by rulez86** [comment]
@milo526 
May you can give me short explanation on how to install a forked version in homebridge, so I could also test the version created from bsr43? Is this possible?

I'm not really experienced in npm installations, using forked versions. Sorry.

I'm running a Debian Server for homebridge, if this information is needed.

Regards,
Rulez86

**Comment by iDreamsX** [comment]
@rulez86
Hi, for my part I directly used the graphical aspect of Hombridge by entering "smartlife" and looking for the one of Bsr43. 
But if you don't have the GUI, here is the line to insert if you don't have the GUI : 

`npm i -g @bsr43/homebridge-tuya-web`

**Comment by rulez86** [comment]
@iDreamsX 
Thanks for the answer. Yes, I have graphic ui and tried this, but it's always just installing version v0.7.1 which is the fork of original package.
When I compare v0.7.1 with the current master, there are 4 changes files where the fix is the in. So I need current master.

Did you made any special thing to force installation of current master?

Did you install bsr43 version in parallel of the milo version, or uninstall milo before? For me it's not working to install current master.

Thanks in advice

**Comment by milo526** [comment]
> @iDreamsX
> Thanks for the answer. Yes, I have graphic ui and tried this, but it's always just installing version v0.7.1 which is the fork of original package.
> When I compare v0.7.1 with the current master, there are 4 changes files where the fix is the in. So I need current master.
> 
> Did you made any special thing to force installation of current master?
> 
> Did you install bsr43 version in parallel of the milo version, or uninstall milo before? For me it's not working to install current master.
> 
> Thanks in advice

They will probably not work together since they use the same platform identifiers.

On the other hand they should also be interchangeable, you should be able to remove my version, install @bSr43 their version and restart homebridge without any problems.

Please note that installing a forked version will prevent you from seeing updates in the homebridge ui of my version.
So please keep an eye out for new version that are released here (or permanently switch to @bSr43 their branch).

**Comment by rulez86** [comment]
@milo526
Thanks for your answer. If that is true and version v0.7.1 from @bsr43 is the version including the fixes, it unfortunately does not work for me. Exactly same issue as in you v0.7.1 or v0.8.0 release.

Is there any possibility to check code or some other idea to make sure, the changes are running in my installation? From my point of view it looks like there is no different from @bsr43 v0.7.1 to @milo526 v0.7.1

Thanks

**Comment by milo526** [comment]
Go to one of the [changed files](https://github.com/milo526/homebridge-tuya-web/pull/159/files) in your `node_modules` directory and see if the files there match "my" repository or the other version.

**Comment by rulez86** [comment]
> changed files

@milo526 thanks for the help, i got the changed files successfully from @bSr43 . So it looks like it simply does not work for me.
This is my curtain switch module, i'm using: [If](https://www.banggood.com/MoesHouse-MS-108-WiFi-Mini-Smart-Curtain-Switch-Module-Roller-Blinds-Shutter-Motor-Smart-Life-Tuya-APP-Remote-Control-Work-with-Alexa-Google-Home-p-1679965.html?utm_source=googleshopping&utm_medium=cpc_organic&gmcCountry=DE&utm_content=minha&utm_campaign=minha-de-en-pc&currency=EUR&cur_warehouse=CN&createTmp=1&utm_source=googleshopping&utm_medium=cpc_bgs&utm_content=frank&utm_campaign=frank-ssc-de-css-all-newcustom-ncv90-0623&ad_id=443949120684&gclid=EAIaIQobChMIsuviuf-i7gIVEKSyCh0u7g7eEAQYASABEgJCTvD_BwE) 

If you are willing to send me (4sure as a private message) your postal address, i can send you a curtain switch? May, then you can find out the problem? It's really comprehensible that this is a problem without having it physically.

@milo526 Or does it also help to give you a Tuya login where the device is added to? 

@iDreamsX 
Which Tuya device are you using?

**Comment by iDreamsX** [comment]
> @iDreamsX
> Which Tuya device are you using?

I only have the switches for the roller shutters from Loratap, they work with SmartLife including Tuya.

**Comment by iDreamsX** [comment]
@milo526 
I have just noticed that in the SmartLife application, when the switch is pressed manually, the status of the virtual one on the application also changes. As if we had done it directly on the application.

This may give you a clue to see if there is a way to know the current status.

**Comment by milo526** [comment]
> @milo526
> I have just noticed that in the SmartLife application, when the switch is pressed manually, the status of the virtual one on the application also changes. As if we had done it directly on the application.
> 
> This may give you a clue to see if there is a way to know the current status.

I get why you would think that and thanks for trying to help and share information.

However, what you observed has very little to do with this plugin.  
This plugin uses a very specific API that has very limited capabilities.  
Therefor what you observe in the Tuya (or Smart-Life) app, almost never has big influence on this plugin.
